core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala

```diff
@@ -0,0 +1,130 @@
+@InternalApi private[akka] object AllPersistenceIdsStage {
+
+  case object Continue
+
+}
+
+/**
+  * INTERNAL API
+  */
+@InternalApi private[akka] class AllPersistenceIdsStage(
+    refreshInterval: Option[FiniteDuration],
+    fetchSize: Int,
+    preparedStatement: PreparedStatement,
+    session: Session)
+    extends GraphStage[SourceShape[String]] {
+
+  import AllPersistenceIdsStage._
+
+  val out: Outlet[String] = Outlet("AllPersistenceIds.out")
+
+  val shape: SourceShape[String] = SourceShape(out)
+
+  def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new TimerGraphStageLogicWithLogging(shape) {
+
+      implicit def executionContext: ExecutionContextExecutor =
+        materializer.executionContext
+
+      private var knownPersistenceIds: Set[String] = Set.empty
+      private var maybeResultSet = Option.empty[ResultSet]
+      private var buffer: Queue[String] = Queue.empty[String]
+
+      private val queryCallback: AsyncCallback[ResultSet] =
+        getAsyncCallback[ResultSet] { rs =>
+          maybeResultSet = Some(rs)
+          val available = rs.getAvailableWithoutFetching
+          for (_ <- 1 to available) {
+            val s = rs.one().getString("persistence_id")
+            if (!knownPersistenceIds.contains(s)) {
+              buffer = buffer.enqueue(s)
+              knownPersistenceIds += s
+            }
+          }
+          pushWhileAvailable()
+          if (refreshInterval.isEmpty && buffer.isEmpty && rs.getExecutionInfo.getPagingState == null) {
+            complete(out)
+          }
+        }
+
+      private def query(): Unit = {
+        val boundStatement = preparedStatement.bind()
+        boundStatement.setFetchSize(fetchSize)
+        session
+          .executeAsync(boundStatement)
+          .asScala
+          .foreach(queryCallback.invoke)
+      }
+
+      private def pushWhileAvailable(): Unit = {
+        while (buffer.nonEmpty && isAvailable(out)) {
+          val (s, newBuffer) = buffer.dequeue
+          buffer = newBuffer
+          push(out, s)
+        }
+      }
+
+      override def preStart(): Unit = {
+        refreshInterval.foreach { interval =>
+          val initial =
+            if (interval >= 2.seconds)
+              ThreadLocalRandom.current().nextLong(interval.toMillis).millis
+            else interval
+          schedulePeriodicallyWithInitialDelay(Continue, initial, interval)
+        }
+      }
+
+      override def onTimer(timerKey: Any): Unit = {
+        timerKey match {
+          case Continue =>
+            query()
+
+          case _ =>
+        }
+      }
+
+      setHandler(
+        out,
+        new OutHandler {
+          def onPull(): Unit = {
+            pushWhileAvailable()
+            if (buffer.isEmpty && isAvailable(out)) {
+              maybeResultSet match {
+                case None =>
+                  query()
+
+                case Some(rs) =>
+                  if (rs.getExecutionInfo.getPagingState != null) {
```

Comment:
I think this should check if the result set is exhausted, if so then kick off another query. If not and getAvailableWithoutFetching it can pull more off, otherwise it should kick off the fetchMoreResults

Reply:
This is now achieved by holding on to the `resultSet` and calling `fetchMoreResults` when demand is signalled while the buffer is empty.